// @tialops/maki/src/dev.js
import { spawn } from 'child_process'
import { createServer } from 'vite'
import { fileURLToPath } from 'url'
import fs from 'fs'
import { getBaseViteConfig } from './vite.js'

export async function dev() {
    if (!fs.existsSync('./scenes')) {
        console.log('No scenes folder found. Are you inside a maki game folder?')
        console.log('Try: maki new my-game && cd my-game && maki dev')
        return
    }

    console.log('🍣 Starting Maki dev server...')

    const devserverPath = fileURLToPath(new URL('../lib/devserver.js', import.meta.url))
    const devserver = spawn(process.execPath, [devserverPath], {
        stdio: 'inherit',
        cwd: process.cwd(),
    })

    const server = await createServer({
        ...getBaseViteConfig(),
        server: {
            port: 5173,
            open: true
        }
    })

    const stop = async (code = 0) => {
        devserver.kill('SIGTERM')
        await server.close()
        process.exit(code)
    }

    devserver.on('exit', async code => {
        await server.close()
        if (code && code !== 0) {
            process.exit(code)
        }
    })

    process.on('SIGINT', () => {
        void stop(0)
    })

    process.on('SIGTERM', () => {
        void stop(0)
    })

    await server.listen()
    console.log('🍣 Game running at http://localhost:5173')
}
