// @tialops/maki/src/MakiPlayer.js
export class MakiPlayer {
    constructor(scene, name = 'lia') {
        this.scene = scene
        this.name = name
        this.sprite = null
        this.keys = null
        this.speed = 160
    }

    preload(config) {
        this.config = config
        const { sprite } = config
        if (!this.scene.textures.exists(this.name)) {
            this.scene.load.spritesheet(this.name, `sprites/${sprite.file}`, {
                frameWidth: sprite.frameWidth,
                frameHeight: sprite.frameHeight
            })
        }
    }

    init() {
        const { sprite } = this.config

        Object.keys(sprite.directions).forEach(dir => {
            const { start, end } = sprite.directions[dir]
            const animKey = `${this.name}-${dir}`
            if (!this.scene.anims.exists(animKey)) {
                this.scene.anims.create({
                    key: animKey,
                    frames: this.scene.anims.generateFrameNumbers(this.name, { start, end }),
                    frameRate: 8,
                    repeat: -1
                })
            }
        })

        this.sprite = this.scene.physics.add.sprite(
            this.scene.scale.width / 2,
            this.scene.scale.height / 2,
            this.name
        )
        this.sprite.setDepth(1)

        this.keys = this.scene.input.keyboard.createCursorKeys()
    }

    move() {
        if (!this.sprite) return

        this.sprite.setVelocity(0)

        if (this.keys.left.isDown) {
            this.sprite.setVelocityX(-this.speed)
            this.sprite.anims.play(`${this.name}-left`, true)
        } else if (this.keys.right.isDown) {
            this.sprite.setVelocityX(this.speed)
            this.sprite.anims.play(`${this.name}-right`, true)
        } else if (this.keys.up.isDown) {
            this.sprite.setVelocityY(-this.speed)
            this.sprite.anims.play(`${this.name}-up`, true)
        } else if (this.keys.down.isDown) {
            this.sprite.setVelocityY(this.speed)
            this.sprite.anims.play(`${this.name}-down`, true)
        } else {
            this.sprite.anims.stop()
        }
    }

    onCollide(other) {}

    onOverlap(other) {}
}
