// @tialops/maki/src/sprite.js
import prompts from 'prompts'
import fs from 'fs'

export async function sprite() {
    if (!fs.existsSync('./assets/sprites')) {
        console.log('🍣 No sprites folder found! Are you inside a maki game folder?')
        return
    }

    const files = fs.readdirSync('./assets/sprites').filter(f => f.endsWith('.png'))

    if (files.length === 0) {
        console.log('🍣 No PNG files found in /sprites! Drop your spritesheet in there first.')
        return
    }

    console.log('🍣 Sprites found:', files.join(', '))

    const base = await prompts([
        {
            type: 'select',
            name: 'file',
            message: 'Which spritesheet?',
            choices: files.map(f => ({ title: f, value: f }))
        },
        {
            type: 'select',
            name: 'layout',
            message: 'How is your spritesheet laid out?',
            choices: [
                { title: 'Rows (each row is a direction)', value: 'row' },
                { title: 'Columns (each column is a direction)', value: 'col' }
            ]
        },
        {
            type: 'number',
            name: 'cols',
            message: 'How many columns?'
        },
        {
            type: 'number',
            name: 'rows',
            message: 'How many rows?'
        },
        {
            type: 'number',
            name: 'frameWidth',
            message: 'Frame width in pixels?'
        },
        {
            type: 'number',
            name: 'frameHeight',
            message: 'Frame height in pixels?'
        }
    ])

    // loop through rows or columns depending on layout
    const count = base.layout === 'row' ? base.rows : base.cols
    const label = base.layout === 'row' ? 'Row' : 'Column'
    const directions = {}

    for (let i = 0; i < count; i++) {
        const { direction } = await prompts({
            type: 'select',
            name: 'direction',
            message: `${label} ${i} direction?`,
            choices: [
                { title: 'down', value: 'down' },
                { title: 'up', value: 'up' },
                { title: 'left', value: 'left' },
                { title: 'right', value: 'right' },
                { title: 'idle', value: 'idle' }
            ]
        })
        directions[direction] = i
    }

    const config = {
        sprite: {
            file: base.file,
            layout: base.layout,
            cols: base.cols,
            rows: base.rows,
            frameWidth: base.frameWidth,
            frameHeight: base.frameHeight,
            directions
        }
    }

    fs.writeFileSync('./maki.config.js', `export default ${JSON.stringify(config, null, 4)}`)
    console.log('🍣 Sprite config saved to maki.config.js!')
    console.log('🍣 Restart maki dev to see your sprite!')
}
