// @tialops/maki/src/tilemap.js
import { spawn } from 'child_process'
import { fileURLToPath } from 'url'
import http from 'http'
import fs from 'fs'
import path from 'path'
import prompts from 'prompts'

function scanAssets(assetsDir) {
    const excluded = new Set(['maps', 'sprites'])
    const results = []
    function scan(dir, rel) {
        let entries
        try { entries = fs.readdirSync(dir, { withFileTypes: true }) } catch { return }
        for (const entry of entries) {
            if (entry.isDirectory()) {
                if (!excluded.has(entry.name)) scan(path.join(dir, entry.name), rel ? `${rel}/${entry.name}` : entry.name)
            } else if (/\.(png|jpg|jpeg|gif|webp)$/i.test(entry.name)) {
                results.push(rel ? `${rel}/${entry.name}` : entry.name)
            }
        }
    }
    scan(assetsDir, '')
    return results
}

function readGameDimensions() {
    try {
        const text = fs.readFileSync(path.join(process.cwd(), 'maki.config.js'), 'utf8')
        const w = text.match(/\bwidth\s*:\s*(\d+)/)
        const h = text.match(/\bheight\s*:\s*(\d+)/)
        return {
            width:  w ? parseInt(w[1]) : 800,
            height: h ? parseInt(h[1]) : 600,
        }
    } catch {
        return { width: 800, height: 600 }
    }
}

const __filename = fileURLToPath(import.meta.url)

function openBrowser(url) {
    const platform = process.platform
    const cmd = platform === 'darwin' ? 'open' : platform === 'win32' ? 'cmd' : 'xdg-open'
    const args = platform === 'win32' ? ['/c', 'start', '', url] : [url]
    spawn(cmd, args, { stdio: 'ignore', detached: true }).unref()
}

function isDevserverRunning() {
    return new Promise(resolve => {
        const req = http.get({ hostname: 'localhost', port: 3001, path: '/', timeout: 500 }, () => {
            req.destroy()
            resolve(true)
        })
        req.on('error', () => resolve(false))
        req.on('timeout', () => { req.destroy(); resolve(false) })
    })
}

function listTilesets() {
    const assetsDir = path.join(process.cwd(), 'assets')
    const results = []
    let topLevel
    try { topLevel = fs.readdirSync(assetsDir, { withFileTypes: true }) } catch { return results }
    // rooms first, then other subdirs (skip sprites and maps)
    const subdirs = topLevel
        .filter(e => e.isDirectory() && !['sprites', 'maps'].includes(e.name))
        .sort((a, b) => (a.name === 'rooms' ? -1 : b.name === 'rooms' ? 1 : a.name.localeCompare(b.name)))
    for (const entry of subdirs) {
        try {
            const files = fs.readdirSync(path.join(assetsDir, entry.name))
            for (const file of files) {
                if (/\.(png|jpg|jpeg|gif|webp)$/i.test(file)) {
                    results.push(`assets/${entry.name}/${file}`)
                }
            }
        } catch { /* skip unreadable subdir */ }
    }
    return results
}

export async function tilemap({ collisionMode = false } = {}) {
    const tilesets = listTilesets()
    // Auto-pick the first available room tileset — can be switched inside the editor
    const defaultTileset = tilesets.find(f => f.includes('rooms/')) || tilesets[0] || 'assets/rooms/room1.png'

    const base = await prompts(
        [
            {
                type: 'select',
                name: 'tileSize',
                message: 'Tile size in pixels?',
                choices: [
                    { title: '8', value: 8 },
                    { title: '16', value: 16 },
                    { title: '32', value: 32 },
                    { title: '64', value: 64 },
                    { title: 'custom', value: 'custom' },
                ],
                initial: 1,
            },
            {
                type: prev => (prev === 'custom' ? 'number' : null),
                name: 'tileSizeCustom',
                message: 'Custom tile size in pixels?',
                validate: v => (v > 0 ? true : 'Must be > 0'),
            },
            {
                type: 'text',
                name: 'mapName',
                message: 'Map name?',
                initial: 'default',
            },
            {
                type: 'confirm',
                name: 'loadExisting',
                message: 'Load existing map to edit? (loads assets/maps/<name>.json)',
                initial: false,
            },
        ],
        { onCancel: () => process.exit(0) }
    )

    base.tileset = defaultTileset

    const tileSize = base.tileSize === 'custom' ? base.tileSizeCustom : base.tileSize

    // Default map dimensions: match the game canvas size from maki.config.js
    const game = readGameDimensions()
    let defaultWidth  = Math.round(game.width  / tileSize)
    let defaultHeight = Math.round(game.height / tileSize)

    // If editing an existing map, use its saved dimensions instead
    if (base.loadExisting) {
        try {
            const mapData = JSON.parse(
                fs.readFileSync(path.join(process.cwd(), 'assets', 'maps', `${base.mapName}.json`), 'utf8')
            )
            if (mapData.mapWidth)  defaultWidth  = mapData.mapWidth
            if (mapData.mapHeight) defaultHeight = mapData.mapHeight
        } catch { /* file missing, use game-canvas defaults */ }
    }

    const dims = await prompts(
        [
            {
                type: 'number',
                name: 'mapWidth',
                message: 'Map width (in tiles)?',
                initial: defaultWidth,
                validate: v => (v > 0 ? true : 'Must be > 0'),
            },
            {
                type: 'number',
                name: 'mapHeight',
                message: 'Map height (in tiles)?',
                initial: defaultHeight,
                validate: v => (v > 0 ? true : 'Must be > 0'),
            },
        ],
        { onCancel: () => process.exit(0) }
    )

    const answers = { ...base, ...dims }

    const params = new URLSearchParams({
        tileset: answers.tileset,
        tileSize,
        mapWidth: answers.mapWidth,
        mapHeight: answers.mapHeight,
        mapName: answers.mapName,
        loadExisting: answers.loadExisting ? '1' : '0',
        mode: collisionMode ? 'collision' : 'paint',
    })

    const url = `http://localhost:3001/tilemap-editor?${params}`

    const running = await isDevserverRunning()
    let devserverProc = null

    if (!running) {
        const devserverPath = fileURLToPath(new URL('../lib/devserver.js', import.meta.url))
        devserverProc = spawn(process.execPath, [devserverPath], {
            stdio: 'inherit',
            cwd: process.cwd(),
        })
        await new Promise(resolve => setTimeout(resolve, 600))
    }

    console.log(`\nOpening map editor at ${url}`)
    openBrowser(url)

    if (devserverProc) {
        process.on('SIGINT', () => { devserverProc.kill('SIGTERM'); process.exit(0) })
        process.on('SIGTERM', () => { devserverProc.kill('SIGTERM'); process.exit(0) })
        devserverProc.on('exit', code => process.exit(code ?? 0))
        await new Promise(() => {})
    }
}
