// @tialops/maki/index.js
import Phaser from 'phaser'
import { MakiPlayer } from './src/MakiPlayer.js'
import { manager } from './lib/manager.js'

export { MakiPlayer, manager }

export class Scene extends Phaser.Scene {
    constructor(config) {
        super(config)
        this._makiPlayers = []
    }

    preload() {
        this.maki = {
            player: (name = 'lia') => {
                const p = new MakiPlayer(this, name)
                p.preload(this._getConfig())
                this._makiPlayers.push(p)
                return p
            },
            move: (player) => {
                player.move()
            }
        }
    }

    create() {
        this._makiPlayers.forEach(p => p.init())
    }

    _getConfig(){
        return{
            sprite:{
                file: 'lia.png',
                layout :'row',
                cols: 24,
                rows: 1,
                frameWidth: 32,
                frameHeight: 64,
                directions: {
                    right: { start: 0, end: 5 },
                    up: { start: 6, end: 11 },
                    left: { start: 12, end: 17 },
                    down: { start: 18, end: 23 }
                }
            }
        }
    }
}
